### PR TITLE
Closes #13690: List all objects to be deleted

### LINKED ISSUE #13690: when deleteing an object in webui. list the related objects that would also be deleted
### NetBox version

v3.6.0

### Feature type

Change to existing functionality

### Proposed functionality

when choosing delete on an object (eg device for this example). the popup that "asks are you sure you want to delete device xxx" should be extended to show the list of related objects that will also be deleted (interfaces, cables, etc)

### Use case

some users are not aware that deleting an object will remove some other related objects.
showing the user what additional objects will deleted would be a fantastic safety feature.

### Database changes

n/a

### External dependencies

n/a

### PR DESCRIPTION
**Pull request by ITJamie**
### Fixes: #13690

List all objects to be deleted when deleting an item.

Currently, the list includes the original item being deleted. we can change that pretty easily but I would like feedback first as to whether the list should be exhaustive or just the additional items (see #13842 as to why including the original item in the list might make sense)

Screenshot 2023-10-20 at 07 52 22
Screenshot 2023-10-20 at 07 53 26

### CODE REVIEW

File: netbox/netbox/views/generic/object_views.py [hunk 1]
```diff
@@ -334,6 +335,18 @@ def get(self, request, *args, **kwargs):
...
         form = ConfirmationForm(initial=request.GET)
 
+        using = router.db_for_write(obj._meta.model)
+        collector = Collector(using=using)
+        collector.collect([obj])
+        deletion_objects = []
+        if collector.instances_with_model():
+            for model, instance in collector.instances_with_model():
+                # we could ignore the instance == obj so that the list doesnt contain itself...
+                deletion_objects.append({
```
Comment (ITJamie) [inline netbox/netbox/views/generic/object_views.py:345]: this append could be gated to not include the original item
```
        if collector.instances_with_model():
            for model, instance in collector.instances_with_model():
                if not instance == obj:
                    deletion_objects.append({
                        "modelname":f"{model.__name__}", 
                        "object": instance,
                    })
```

**Comment by abhi1693** [comment]
I like the general idea but this will turn out to be a bad ux if the similar objects are listed in the new line because it may go beyond the actual page making a huge scroll. I'd prefer the similar objects be comma separated instead.

File: netbox/netbox/views/generic/object_views.py [hunk 2]
```diff
@@ -334,6 +335,18 @@ def get(self, request, *args, **kwargs):
...
+                    "modelname":f"{model.__name__}",
```
Comment (abhi1693) [inline netbox/netbox/views/generic/object_views.py:346]: What's the need of the f-string here? Isn't the name already a string?

**Review by abhi1693 — CHANGES_REQUESTED** [review]

**Comment by ITJamie** [comment]
> I like the general idea but this will turn out to be a bad ux if the similar objects are listed in the new line because it may go beyond the actual page making a huge scroll. I'd prefer the similar objects be comma separated instead.

I can absolutely refactor for that. For switches it makes a lot of sense as there can be a lot of ports and make the list long. 

The one advantage to that is that it really reinforces just how much impact the current deletion will have by having to scroll to the delete button 😂

**Comment by abhi1693** [comment]
I just noticed this is to be released in v3.7, so it should go into the feature branch instead.

**Comment by ITJamie** [comment]
Screenshot 2023-10-20 at 12 04 29
@abhi1693 something more like this?

File: netbox/netbox/views/generic/object_views.py [hunk 3]
```diff
@@ -334,6 +335,18 @@ def get(self, request, *args, **kwargs):
         obj = self.get_object(**kwargs)
         form = ConfirmationForm(initial=request.GET)
 
+        using = router.db_for_write(obj._meta.model)
+        collector = Collector(using=using)
+        collector.collect([obj])
+        deletion_objects = []
+        if collector.instances_with_model():
```
[the remark this answers is not in the document]
Comment (ITJamie) [reply netbox/netbox/views/generic/object_views.py]: removed the unneeded if statement

File: netbox/netbox/views/generic/object_views.py (hunk 2 above)
Comment (ITJamie) [reply netbox/netbox/views/generic/object_views.py]: cleaned up. was a leftover from my first pass

**Comment by ITJamie** [comment]
@kkthxbye-code re the upvotes and downvotes. Is there anything specific you would like me to adjust regards the ui view change from the first screenshots to the adjusted screenshots based on @abhi1693 feedback?

**Comment by kkthxbye-code** [comment]
> @kkthxbye-code re the upvotes and downvotes. Is there anything specific you would like me to adjust regards the ui view change from the first screenshots to the adjusted screenshots based on @abhi1693 feedback?

I think your initial version was much more readble than the version abhi suggested. Showing a large number of varying length list items as comma seperated tends to look very messy.

If the page size is an issue (I personally don't think it is), maybe a scrolling container could solve it, or truncating the output with a "show more" button that extends it.

Also the trailing comma should be removed if the second version is chosen.

File: netbox/templates/htmx/delete_form.html [hunk 4]
```diff
@@ -12,6 +12,18 @@ <h5 class="modal-title">{% trans "Confirm Deletion" %}</h5>
...
+    {% if related_objects %}
+    <p>
+      This will cause deletion of the following items: <br>
+      {% for model,objects in related_objects.items %} 
+        - {{ model }}:
+        {% for object in objects %}
+          {{ object | linkify }}{% if not forloop.last %}, {% endif %}
+        {% endfor %}
+        <br>
+      {% endfor %}
```
Comment (jeremystretch) [inline netbox/templates/htmx/delete_form.html:24]: IMO we should present the list of dependent objects in a scrollable list. Bootstrap's [accordion component](https://getbootstrap.com/docs/5.1/components/accordion/) could work very well here.

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/templates/htmx/delete_form.html (hunk 4 above)
Comment (ITJamie) [reply netbox/templates/htmx/delete_form.html:24]: Screenshot 2023-10-31 at 21 14 27
Screenshot 2023-10-31 at 21 14 32

happier with this? (im using verbose_name and verbose_name_plural as needed for the accordion headers, some are capitalised, some are not)

**Comment by ITJamie** [comment]
Current version:
- using bootstrap accordian
- titles are now being passed in + using plurals 
- added `_get_dependent_objects` to `ObjectDeleteView` class
- using 
Screenshot 2023-10-31 at 21 50 47
Screenshot 2023-10-31 at 21 50 58

**Review by jeremystretch — APPROVED** [review]
Nice work @ITJamie!

**Review by jeremystretch — APPROVED** [review]

**Pull request merged**